// appium-espresso-driver/build/lib/commands/general.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumSupport = require("appium-support");

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

function assertRequiredOptions(options, requiredOptionNames) {
  if (!_lodash.default.isArray(requiredOptionNames)) {
    requiredOptionNames = [requiredOptionNames];
  }

  const presentOptionNames = _lodash.default.keys(options);

  const missingOptionNames = _lodash.default.difference(requiredOptionNames, presentOptionNames);

  if (_lodash.default.isEmpty(missingOptionNames)) {
    return options;
  }

  throw new Error(`The following options are required: ${JSON.stringify(missingOptionNames)}. ` + `You have only provided: ${JSON.stringify(presentOptionNames)}`);
}

commands.mobilePerformEditorAction = function () {
  var _ref = (0, _asyncToGenerator2.default)(function* (opts = {}) {
    const _assertRequiredOption = assertRequiredOptions(opts, ['action']),
          action = _assertRequiredOption.action;

    return yield this.espresso.jwproxy.command('/appium/device/perform_editor_action', 'POST', {
      action
    });
  });

  return function () {
    return _ref.apply(this, arguments);
  };
}();

commands.mobileSwipe = function () {
  var _ref2 = (0, _asyncToGenerator2.default)(function* (opts = {}) {
    const _assertRequiredOption2 = assertRequiredOptions(opts, ['direction', 'element']),
          direction = _assertRequiredOption2.direction,
          element = _assertRequiredOption2.element;

    return yield this.espresso.jwproxy.command(`/appium/execute_mobile/${_appiumSupport.util.unwrapElement(element)}/swipe`, 'POST', {
      direction
    });
  });

  return function () {
    return _ref2.apply(this, arguments);
  };
}();

commands.mobileGetDeviceInfo = (0, _asyncToGenerator2.default)(function* () {
  return yield this.espresso.jwproxy.command('/appium/device/info', 'GET');
});

helpers.suspendChromedriverProxy = function () {
  this.chromedriver = null;
  this.proxyReqRes = this.espresso.proxyReqRes.bind(this.espresso);
  this.jwpProxyActive = true;
};

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
