// appium-xcuitest-driver/build/lib/commands/cookies.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _logger = _interopRequireDefault(require("../logger"));

var _appiumBaseDriver = require("appium-base-driver");

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

commands.getCookies = function () {
  var _getCookies = (0, _asyncToGenerator2.default)(function* () {
    if (!this.isWebContext()) {
      throw new _appiumBaseDriver.errors.NotImplementedError();
    }

    const cookies = (yield this.remote.getCookies()) || {
      cookies: []
    };
    return cookies.cookies.map(function (cookie) {
      return Object.assign({}, cookie, {
        value: decodeURI(cookie.value)
      });
    });
  });

  return function getCookies() {
    return _getCookies.apply(this, arguments);
  };
}();

commands.deleteCookie = function () {
  var _deleteCookie2 = (0, _asyncToGenerator2.default)(function* (cookieName) {
    if (!this.isWebContext()) {
      throw new _appiumBaseDriver.errors.NotImplementedError();
    }

    const cookies = yield this.getCookies();
    const cookie = cookies.find(cookie => cookie.name === cookieName);

    if (!cookie) {
      _logger.default.debug(`Cookie '${cookieName}' not found. Ignoring.`);

      return true;
    }

    yield this._deleteCookie(cookie);
    return true;
  });

  return function deleteCookie(_x) {
    return _deleteCookie2.apply(this, arguments);
  };
}();

commands.deleteCookies = function () {
  var _deleteCookies = (0, _asyncToGenerator2.default)(function* () {
    if (!this.isWebContext()) {
      throw new _appiumBaseDriver.errors.NotImplementedError();
    }

    const cookies = yield this.getCookies();
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = cookies[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        const cookie = _step.value;
        yield this._deleteCookie(cookie);
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    return true;
  });

  return function deleteCookies() {
    return _deleteCookies.apply(this, arguments);
  };
}();

helpers._deleteCookie = function () {
  var _deleteCookie3 = (0, _asyncToGenerator2.default)(function* (cookie) {
    const url = `http${cookie.secure ? 's' : ''}://${cookie.domain}${cookie.path}`;
    return yield this.remote.deleteCookie(cookie.name, url);
  });

  return function _deleteCookie(_x2) {
    return _deleteCookie3.apply(this, arguments);
  };
}();

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
