// appium-xcuitest-driver/build/lib/commands/execute.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumIosDriver = require("appium-ios-driver");

var _appiumBaseDriver = require("appium-base-driver");

let extensions = {};
Object.assign(extensions, _appiumIosDriver.iosCommands.execute);
const iosExecute = extensions.execute;

extensions.execute = function () {
  var _ref = (0, _asyncToGenerator2.default)(function* (script, args) {
    if (!script.match(/^mobile:/) && !this.isWebContext()) {
      throw new _appiumBaseDriver.errors.NotImplementedError();
    }

    return yield iosExecute.call(this, script, args);
  });

  return function (_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

const iosExecuteAsync = extensions.executeAsync;

extensions.executeAsync = function () {
  var _ref2 = (0, _asyncToGenerator2.default)(function* (script, args, sessionId) {
    if (!this.isWebContext()) {
      throw new _appiumBaseDriver.errors.NotImplementedError();
    }

    return yield iosExecuteAsync.call(this, script, args, sessionId);
  });

  return function (_x3, _x4, _x5) {
    return _ref2.apply(this, arguments);
  };
}();

extensions.executeMobile = function () {
  var _ref3 = (0, _asyncToGenerator2.default)(function* (mobileCommand, opts = {}) {
    const commandMap = {
      scroll: 'mobileScroll',
      swipe: 'mobileSwipe',
      pinch: 'mobilePinch',
      doubleTap: 'mobileDoubleTap',
      twoFingerTap: 'mobileTwoFingerTap',
      touchAndHold: 'mobileTouchAndHold',
      tap: 'mobileTap',
      dragFromToForDuration: 'mobileDragFromToForDuration',
      selectPickerWheelValue: 'mobileSelectPickerWheelValue',
      alert: 'mobileHandleAlert',
      setPasteboard: 'mobileSetPasteboard',
      getPasteboard: 'mobileGetPasteboard',
      source: 'mobileGetSource',
      getContexts: 'mobileGetContexts',
      installApp: 'mobileInstallApp',
      isAppInstalled: 'mobileIsAppInstalled',
      removeApp: 'mobileRemoveApp',
      launchApp: 'mobileLaunchApp',
      terminateApp: 'mobileTerminateApp',
      queryAppState: 'mobileQueryAppState',
      activateApp: 'mobileActivateApp',
      viewportScreenshot: 'getViewportScreenshot',
      startPerfRecord: 'mobileStartPerfRecord',
      stopPerfRecord: 'mobileStopPerfRecord',
      installCertificate: 'mobileInstallCertificate',
      startLogsBroadcast: 'mobileStartLogsBroadcast',
      stopLogsBroadcast: 'mobileStopLogsBroadcast',
      batteryInfo: 'mobileGetBatteryInfo',
      pressButton: 'mobilePressButton',
      enrollBiometric: 'mobileEnrollBiometric',
      sendBiometricMatch: 'mobileSendBiometricMatch',
      isBiometricEnrolled: 'mobileIsBiometricEnrolled',
      clearKeychains: 'mobileClearKeychains',
      getPermission: 'mobileGetPermission'
    };

    if (!_lodash.default.has(commandMap, mobileCommand)) {
      throw new _appiumBaseDriver.errors.UnknownCommandError(`Unknown mobile command "${mobileCommand}". Only ${_lodash.default.keys(commandMap)} commands are supported.`);
    }

    return yield this[commandMap[mobileCommand]](opts);
  });

  return function (_x6) {
    return _ref3.apply(this, arguments);
  };
}();

var _default = extensions;
exports.default = _default;require('source-map-support').install();
