// appium-xcuitest-driver/build/lib/commands/find.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumBaseDriver = require("appium-base-driver");

var _appiumSupport = require("appium-support");

var _logger = _interopRequireDefault(require("../logger"));

const MAGIC_FIRST_VIS_CHILD_SEL = /\/\*\[@firstVisible\s*=\s*('|")true\1\]/;
const MAGIC_SCROLLABLE_SEL = /\/\/\*\[@scrollable\s*=\s*('|")true\1\]/;
const WDA_CLASS_CHAIN_STRATEGY = 'class chain';
let helpers = {},
    commands = {},
    extensions = {};
exports.commands = commands;
exports.helpers = helpers;

helpers.findElOrEls = function () {
  var _ref = (0, _asyncToGenerator2.default)(function* (strategy, selector, mult, context) {
    if (this.isWebview()) {
      return yield this.findWebElementOrElements(strategy, selector, mult, context);
    } else {
      return yield this.findNativeElementOrElements(strategy, selector, mult, context);
    }
  });

  return function (_x, _x2, _x3, _x4) {
    return _ref.apply(this, arguments);
  };
}();

helpers.findNativeElementOrElements = function () {
  var _ref2 = (0, _asyncToGenerator2.default)(function* (strategy, selector, mult, context) {
    const initSelector = selector;
    let rewroteSelector = false;

    if (strategy === '-ios predicate string') {
      strategy = 'predicate string';
    } else if (strategy === '-ios class chain') {
      strategy = WDA_CLASS_CHAIN_STRATEGY;
    }

    function stripViewFromSelector(selector) {
      let keepView = ['XCUIElementTypeScrollView', 'XCUIElementTypeCollectionView', 'XCUIElementTypeTextView', 'XCUIElementTypeWebView'].indexOf(selector) >= 0;

      if (!keepView && selector.indexOf('View') === selector.length - 4) {
        return selector.substr(0, selector.length - 4);
      } else {
        return selector;
      }
    }

    if (strategy === 'class name') {
      if (selector.indexOf('UIA') === 0) {
        selector = selector.substring(3);
      }

      if (selector.indexOf('XCUIElementType') !== 0) {
        selector = stripViewFromSelector(`XCUIElementType${selector}`);
        rewroteSelector = true;
      }
    }

    if (strategy === 'xpath' && MAGIC_FIRST_VIS_CHILD_SEL.test(selector)) {
      return yield this.getFirstVisibleChild(mult, context);
    } else if (strategy === 'xpath' && MAGIC_SCROLLABLE_SEL.test(selector)) {
      var _rewriteMagicScrollab = rewriteMagicScrollable(mult);

      var _rewriteMagicScrollab2 = (0, _slicedToArray2.default)(_rewriteMagicScrollab, 2);

      strategy = _rewriteMagicScrollab2[0];
      selector = _rewriteMagicScrollab2[1];
    } else if (strategy === 'xpath') {
      selector = selector.replace(/(^|\/)(UIA)([^[/]+)/g, (str, g1, g2, g3) => {
        rewroteSelector = true;
        return g1 + stripViewFromSelector(`XCUIElementType${g3}`);
      });
    }

    if (rewroteSelector) {
      _logger.default.info(`Rewrote incoming selector from '${initSelector}' to ` + `'${selector}' to match XCUI type. You should consider ` + `updating your tests to use the new selectors directly`);
    }

    return yield this.doNativeFind(strategy, selector, mult, context);
  });

  return function (_x5, _x6, _x7, _x8) {
    return _ref2.apply(this, arguments);
  };
}();

helpers.doNativeFind = function () {
  var _ref3 = (0, _asyncToGenerator2.default)(function* (strategy, selector, mult, context) {
    var _this = this;

    context = _appiumSupport.util.unwrapElement(context);
    let endpoint = `/element${context ? `/${context}/element` : ''}${mult ? 's' : ''}`;
    let body = {
      using: strategy,
      value: selector
    };
    let method = 'POST';
    let els;

    try {
      yield this.implicitWaitForCondition((0, _asyncToGenerator2.default)(function* () {
        try {
          els = yield _this.proxyCommand(endpoint, method, body);

          if (mult) {
            return els && els.length;
          } else {
            return !els.status || els.status === 0;
          }
        } catch (err) {
          els = undefined;
          return false;
        }
      }));
    } catch (err) {
      if (err.message && err.message.match(/Condition unmet/)) {
        els = [];
      } else {
        throw err;
      }
    }

    if (mult) {
      return els;
    } else {
      if (!els || _lodash.default.size(els) === 0) {
        throw new _appiumBaseDriver.errors.NoSuchElementError();
      }

      return els;
    }
  });

  return function (_x9, _x10, _x11, _x12) {
    return _ref3.apply(this, arguments);
  };
}();

helpers.getFirstVisibleChild = function () {
  var _ref5 = (0, _asyncToGenerator2.default)(function* (mult, context) {
    _logger.default.info(`Getting first visible child`);

    if (mult) {
      throw new Error("Cannot get multiple first visible children!");
    }

    if (!context) {
      throw new Error("Cannot get first visible child without a context element");
    }

    let index = 1;

    while (true) {
      const strategy = WDA_CLASS_CHAIN_STRATEGY;
      const selector = `*[${index}]`;
      const nthChild = yield this.doNativeFind(strategy, selector, false, context);
      const visible = yield this.getAttribute('visible', nthChild);

      if (visible === "true") {
        _logger.default.info(`Found first visible child at position ${index}`);

        return nthChild;
      }

      index++;
    }
  });

  return function (_x13, _x14) {
    return _ref5.apply(this, arguments);
  };
}();

function rewriteMagicScrollable(mult) {
  const types = ['ScrollView', 'Table', 'CollectionView', 'WebView'].map(t => `XCUIElementType${t}`);
  const pred = types.map(t => `type == "${t}"`).join(" OR ");
  const strategy = WDA_CLASS_CHAIN_STRATEGY;
  let selector = '**/*[`' + pred + '`]';

  if (!mult) {
    selector += '[1]';
  }

  _logger.default.info("Rewrote request for scrollable descendants to class chain " + `format with selector '${selector}'`);

  return [strategy, selector];
}

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
