// appium-doctor/build/lib/demo.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FileCheck = exports.DirCheck = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _utils = require("./utils");

var _appiumSupport = require("appium-support");

var _teen_process = require("teen_process");

var _doctor = require("./doctor");

var _logger = _interopRequireDefault(require("./logger"));

var _prompt = require("./prompt");

let checks = [];

class DirCheck extends _doctor.DoctorCheck {
  constructor(path) {
    super({
      autofix: false
    });
    this.path = path;
  }

  diagnose() {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (!(yield _appiumSupport.fs.exists(_this.path))) {
        return (0, _utils.nok)(`Could NOT find directory at '${_this.path}'!`);
      }

      let stats = yield _appiumSupport.fs.lstat(_this.path);
      return stats.isDirectory() ? (0, _utils.ok)(`Found directory at: ${_this.path}`) : (0, _utils.nok)(`'${_this.path}' is NOT a directory!`);
    })();
  }

  fix() {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      return `Manually create a directory at: ${_this2.path}`;
    })();
  }

}

exports.DirCheck = DirCheck;
checks.push(new DirCheck('/tmp/appium-doctor'));
checks.push(new DirCheck('/tmp/appium-doctor/demo'));

class FileCheck extends _doctor.DoctorCheck {
  constructor(path) {
    super({
      autofix: true
    });
    this.path = path;
  }

  diagnose() {
    var _this3 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      return (yield _appiumSupport.fs.exists(_this3.path)) ? (0, _utils.ok)(`Found file at: ${_this3.path}`) : (0, _utils.nok)(`Could NOT find file at '${_this3.path}'!`);
    })();
  }

  fix() {
    var _this4 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _logger.default.info(`The following command need be executed: touch '${_this4.path}'`);

      let yesno = yield (0, _prompt.fixIt)();

      if (yesno === 'yes') {
        yield (0, _teen_process.exec)('touch', [_this4.path]);
      } else {
        _logger.default.info(`Skipping you will need to touch '${_this4.path}' manually.`);

        throw new _doctor.FixSkippedError('bbb');
      }
    })();
  }

}

exports.FileCheck = FileCheck;
checks.push(new FileCheck('/tmp/appium-doctor/demo/apple.fruit'));
checks.push(new FileCheck('/tmp/appium-doctor/demo/pear.fruit'));
checks.push(new FileCheck('/tmp/appium-doctor/demo/orange.fruit'));
var _default = checks;
exports.default = _default;require('source-map-support').install();
