// appium-doctor/build/lib/dev.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.AndroidSdkExists = exports.BinaryIsInPathCheck = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _teen_process = require("teen_process");

var _doctor = require("./doctor");

var _utils = require("./utils");

var _appiumSupport = require("appium-support");

var _path = _interopRequireDefault(require("path"));

var _os = require("os");

let checks = [];

class BinaryIsInPathCheck extends _doctor.DoctorCheck {
  constructor(binary) {
    super();
    this.binary = binary;
  }

  diagnose() {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let resolvedPath;

      try {
        let executable = _appiumSupport.system.isWindows() ? 'where' : 'which';

        let _ref = yield (0, _teen_process.exec)(executable, [_this.binary]),
            stdout = _ref.stdout;

        if (stdout.match(/not found/gi)) {
          throw new Error('Not Found');
        }

        resolvedPath = _appiumSupport.system.isWindows() ? stdout.split(_os.EOL)[0] : stdout.replace(_os.EOL, '');
      } catch (err) {
        return (0, _utils.nok)(`${_this.binary} is MISSING in PATH!`);
      }

      return (yield _appiumSupport.fs.exists(resolvedPath)) ? (0, _utils.ok)(`${_this.binary} was found at ${resolvedPath}`) : (0, _utils.nok)(`${_this.binary} was found in PATH at '${resolvedPath}', but this is NOT a valid path!`);
    })();
  }

  fix() {
    return `Manually install the ${this.binary} binary and add it to PATH.`;
  }

}

exports.BinaryIsInPathCheck = BinaryIsInPathCheck;
checks.push(new BinaryIsInPathCheck(_appiumSupport.system.isWindows() ? 'mvn.bat' : 'mvn'));
checks.push(new BinaryIsInPathCheck(_appiumSupport.system.isWindows() ? 'ant.bat' : 'ant'));
checks.push(new BinaryIsInPathCheck(_appiumSupport.system.isWindows() ? 'adb.exe' : 'adb'));

class AndroidSdkExists extends _doctor.DoctorCheck {
  constructor(sdk) {
    super();
    this.sdk = sdk;
  }

  diagnose() {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (typeof process.env.ANDROID_HOME === 'undefined') {
        return (0, _utils.nok)(`${_this2.sdk} could not be found because ANDROID_HOME is NOT set!`);
      }

      let sdkPath = _path.default.resolve(process.env.ANDROID_HOME, _path.default.join("platforms", _this2.sdk));

      return (yield _appiumSupport.fs.exists(sdkPath)) ? (0, _utils.ok)(`${_this2.sdk} was found at: ${sdkPath}`) : (0, _utils.nok)(`${_this2.sdk} could NOT be found at '${sdkPath}'!`);
    })();
  }

  fix() {
    if (typeof process.env.ANDROID_HOME === "undefined") {
      return 'Manually configure ANDROID_HOME.';
    }

    return `Manually install the ${this.sdk} sdk.`;
  }

}

exports.AndroidSdkExists = AndroidSdkExists;
checks.push(new AndroidSdkExists('android-16'));
checks.push(new AndroidSdkExists('android-19'));
var _default = checks;
exports.default = _default;require('source-map-support').install();
