// appium-doctor/build/lib/prompt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.configure = configure;
exports.fixIt = fixIt;
exports.clear = clear;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _utils = require("./utils");

let persistentResponse;
const fixItQuestion = {
  type: 'list',
  name: 'confirmation',
  message: 'Fix it:',
  choices: ['yes', 'no', 'always', 'never'],

  filter(val) {
    return val.toLowerCase();
  }

};

function configure(opts) {
  if (opts.yes) {
    persistentResponse = 'yes';
  }

  if (opts.no) {
    persistentResponse = 'no';
  }
}

function clear() {
  persistentResponse = undefined;
}

function fixIt() {
  return _fixIt.apply(this, arguments);
}

function _fixIt() {
  _fixIt = (0, _asyncToGenerator2.default)(function* () {
    if (persistentResponse) {
      return persistentResponse;
    }

    let resp = yield _utils.inquirer.prompt(fixItQuestion);
    persistentResponse = resp.confirmation === 'always' ? 'yes' : persistentResponse;
    persistentResponse = resp.confirmation === 'never' ? 'no' : persistentResponse;
    return persistentResponse || resp.confirmation;
  });
  return _fixIt.apply(this, arguments);
}require('source-map-support').install();
