// appium-xcuitest-driver/build/lib/device-log/ios-crash-log.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.IOSCrashLog = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _appiumIosDriver = require("appium-ios-driver");

var _appiumSupport = require("appium-support");

var _bluebird = _interopRequireDefault(require("bluebird"));

var _logger = _interopRequireDefault(require("../logger"));

var _teen_process = require("teen_process");

var _path = _interopRequireDefault(require("path"));

class IOSCrashLog extends _appiumIosDriver.IOSCrashLog {
  constructor(opts = {}) {
    super(opts.udid ? _path.default.resolve(process.env.HOME, 'Library', 'Logs', 'CrashReporter', 'MobileDevice') : _path.default.resolve(process.env.HOME, 'Library', 'Logs', 'DiagnosticReports'));
    this.udid = opts.udid;
    this.phoneName = null;
    this.sim = opts.sim;
  }

  getCrashes() {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let crashLogsRoot = _this.logDir;

      if (_this.udid) {
        if (!_this.phoneName) {
          try {
            const _ref = yield (0, _teen_process.exec)('idevicename', ['-u', _this.udid]),
                  stdout = _ref.stdout;

            _this.phoneName = stdout.trim();
          } catch (e) {
            _logger.default.warn(`Cannot get the name of the crashes folder for the device with udid '${_this.udid}'. ` + `Original error: ${e.message}`);

            return [];
          }
        }

        if (_this.phoneName) {
          crashLogsRoot = _path.default.resolve(crashLogsRoot, _this.phoneName);
        }
      }

      if (!(yield _appiumSupport.fs.exists(crashLogsRoot))) {
        _logger.default.debug(`Crash reports root '${crashLogsRoot}' does not exist. Got nothing to gather.`);

        return [];
      }

      const foundFiles = yield _appiumSupport.fs.glob(`${crashLogsRoot}/**/*.crash`);

      if (_this.udid) {
        return foundFiles;
      }

      return yield _bluebird.default.filter(foundFiles, function () {
        var _ref2 = (0, _asyncToGenerator2.default)(function* (x) {
          try {
            const content = yield _appiumSupport.fs.readFile(x, 'utf8');
            return content.toUpperCase().includes(_this.sim.udid.toUpperCase());
          } catch (err) {
            return false;
          }
        });

        return function (_x) {
          return _ref2.apply(this, arguments);
        };
      }());
    })();
  }

  filesToJSON(paths) {
    return (0, _asyncToGenerator2.default)(function* () {
      return yield _bluebird.default.map(paths, function () {
        var _ref3 = (0, _asyncToGenerator2.default)(function* (fullPath) {
          const stat = yield _appiumSupport.fs.stat(fullPath);
          return {
            timestamp: stat.ctime.getTime(),
            level: 'ALL',
            message: yield _appiumSupport.fs.readFile(fullPath, 'utf8')
          };
        });

        return function (_x2) {
          return _ref3.apply(this, arguments);
        };
      }());
    })();
  }

}

exports.IOSCrashLog = IOSCrashLog;
var _default = IOSCrashLog;
exports.default = _default;require('source-map-support').install();
