// appium-xcuitest-driver/build/lib/device-log/rotating-log.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.MAX_LOG_ENTRIES_COUNT = exports.RotatingLog = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumSupport = require("appium-support");

const MAX_LOG_ENTRIES_COUNT = 10000;
exports.MAX_LOG_ENTRIES_COUNT = MAX_LOG_ENTRIES_COUNT;

class RotatingLog {
  constructor(showLogs = false, label = 'Log Label') {
    this.log = _appiumSupport.logger.getLogger(label);
    this.showLogs = showLogs;
    this.logs = [];
    this.logIdxSinceLastRequest = 0;
    this.isCapturing = false;
  }

  startCapture() {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _this.isCapturing = true;
    })();
  }

  stopCapture() {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _this2.isCapturing = false;
    })();
  }

  addLogLine() {}

  getLogs() {
    var _this3 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (_this3.logs.length && _this3.logIdxSinceLastRequest < _this3.logs.length) {
        let result = _this3.logs;

        if (_this3.logIdxSinceLastRequest > 0) {
          result = result.slice(_this3.logIdxSinceLastRequest);
        }

        _this3.logIdxSinceLastRequest = _this3.logs.length;
        return result;
      }

      return [];
    })();
  }

  getAllLogs() {
    var _this4 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      return _lodash.default.clone(_this4.logs);
    })();
  }

  get logs() {
    if (!this._logs) {
      this.logs = [];
    }

    return this._logs;
  }

  set logs(logs) {
    this._logs = logs;
  }

}

exports.RotatingLog = RotatingLog;
var _default = RotatingLog;
exports.default = _default;require('source-map-support').install();
