// appium-youiengine-driver/build/lib/driver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.YouiEngineDriver = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _appiumBaseDriver = require("appium-base-driver");

var _desiredCaps = _interopRequireDefault(require("./desired-caps"));

var _logger = _interopRequireDefault(require("./logger"));

var _index = _interopRequireDefault(require("./commands/index"));

var _lodash = _interopRequireDefault(require("lodash"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _asyncbox = require("asyncbox");

var _appiumAndroidDriver = _interopRequireDefault(require("appium-android-driver"));

var _appiumIosDriver = _interopRequireDefault(require("appium-ios-driver"));

var _appiumXcuitestDriver = _interopRequireDefault(require("appium-xcuitest-driver"));

var _appiumMacDriver = _interopRequireDefault(require("appium-mac-driver"));

const TO_PROXY_COMMON = ['background', 'closeApp', 'getLog', 'getLogTypes', 'getOrientation', 'getStrings', 'launchApp', 'lock', 'removeApp', 'setOrientation'];
const TO_PROXY_IOS_ONLY = ['mobileShake'];
const TO_PROXY_ANDROID_ONLY = ['getNetworkConnection', 'isAppInstalled', 'isLocked', 'longPressKeyCode', 'pressKeyCode', 'setNetworkConnection', 'toggleLocationServices', 'unlock'];
const TO_PROXY_IOS = TO_PROXY_IOS_ONLY.concat(TO_PROXY_COMMON);
const TO_PROXY_ANDROID = TO_PROXY_ANDROID_ONLY.concat(TO_PROXY_COMMON);
const TO_PROXY_MAC = TO_PROXY_COMMON;
const MAX_RETRY_COUNT = 10;
const RETRY_BACKOFF = 3000;

class YouiEngineDriver extends _appiumBaseDriver.BaseDriver {
  resetYouiEngine() {
    this.ready = false;
    this.socket = null;
    this.locatorStrategies = ['id', 'name', 'class name', 'accessibility id'];
    this.proxydriver = null;
    this.proxyAllowList = '';
  }

  constructor(opts, shouldValidateCaps) {
    super(opts, shouldValidateCaps);
    this.desiredCapConstraints = _desiredCaps.default;
    this.settings = new _appiumBaseDriver.DeviceSettings({
      'TimeDilation': 1,
      'SourceTreeFilter': ''
    }, this.onSettingsUpdate.bind(this));
    this.resetYouiEngine();
  }

  validateLocatorStrategy(strategy) {
    super.validateLocatorStrategy(strategy, false);
  }

  createSession(caps) {
    var _this = this,
        _superprop_callCreateSession = (..._args) => super.createSession(..._args);

    return (0, _asyncToGenerator2.default)(function* () {
      try {
        let _ref = yield _superprop_callCreateSession(caps),
            _ref2 = (0, _slicedToArray2.default)(_ref, 1),
            sessionId = _ref2[0];

        if (caps.platformName !== null) {
          let appPlatform = caps.platformName.toLowerCase();

          if (appPlatform === "ios") {
            yield _this.startIOSSession(caps);
          } else if (appPlatform === "android") {
            yield _this.startAndroidSession(caps);
          } else if (appPlatform === "mac") {
            yield _this.startMacSession(caps);
          } else if (appPlatform === "yimac") {
            _this.startYIMacSession(caps);
          } else if (appPlatform === "yiroku") {
            _this.startYIRokuSession(caps);
          } else if (appPlatform === "yitvos") {
            _this.startYITVOSSession(caps);
          }
        }

        yield _this.connectSocket();
        return [sessionId, _this.opts];
      } catch (e) {
        yield _this.deleteSession();
        throw e;
      }
    })();
  }

  onSettingsUpdate(key, value) {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (key === "TimeDilation") {
        yield _this2.setTimeDilation(value);
      } else if (key === "SourceTreeFilter") {
        yield _this2.setSourceTreeFilter(value);
      }
    })();
  }

  stop() {
    var _this3 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _this3.ready = false;
    })();
  }

  deleteSession() {
    var _this4 = this,
        _superprop_callDeleteSession = (..._args2) => super.deleteSession(..._args2);

    return (0, _asyncToGenerator2.default)(function* () {
      _logger.default.debug("Deleting YouiEngine session");

      if (_this4.caps.platformName !== null) {
        let appPlatform = _this4.caps.platformName.toLowerCase();

        if (appPlatform === "yimac") {
          _this4.endYIMacSession(_this4.caps);
        } else if (appPlatform === "yitvos") {
          _this4.endYITVOSSession(_this4.caps);
        }
      }

      if (_this4.proxydriver !== null) {
        yield _this4.proxydriver.deleteSession();
      }

      yield _superprop_callDeleteSession();
      yield _this4.stop();
    })();
  }

  driverShouldDoProxyCmd(command) {
    if (!this.proxydriver) {
      return false;
    }

    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = this.proxyAllowList[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        let allowedCommand = _step.value;

        if (allowedCommand === command) {
          return true;
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    return false;
  }

  executeCommand(cmd, ...args) {
    var _this5 = this,
        _superprop_callExecuteCommand = (..._args3) => super.executeCommand(..._args3);

    return (0, _asyncToGenerator2.default)(function* () {
      if (cmd === 'receiveAsyncResponse') {
        _logger.default.debug(`Executing YouiEngineDriver response '${cmd}'`);

        return yield _this5.receiveAsyncResponse(...args);
      } else if (_this5.ready) {
        if (_this5.driverShouldDoProxyCmd(cmd)) {
          _logger.default.debug(`Executing proxied WebDriver command '${cmd}'`);

          _this5.clearNewCommandTimeout();

          let result = _this5.proxydriver.executeCommand(cmd, ...args);

          _this5.startNewCommandTimeout(cmd);

          return result;
        } else {
          _logger.default.debug(`Executing YouiEngine WebDriver command '${cmd}'`);

          return _superprop_callExecuteCommand(cmd, ...args);
        }
      } else {
        _logger.default.debug(`Command Error '${cmd}'`);

        throw new _appiumBaseDriver.errors.NoSuchDriverError(`Driver is not ready, cannot execute ${cmd}.`);
      }
    })();
  }

  validateDesiredCaps(caps) {
    let res = super.validateDesiredCaps(caps);

    if (!res) {
      return res;
    }

    if (caps.platformName.toLowerCase() !== 'noproxy') {
      if (!caps.youiEngineAppAddress) {
        let msg = 'The desired capabilities must include youiEngineAppAddress';

        _logger.default.errorAndThrow(msg);
      }

      if (!caps.app) {
        let msg = 'The desired capabilities must include app';

        _logger.default.errorAndThrow(msg);
      }

      if (caps.deviceName.toLowerCase() === 'android') {
        if (!caps.avd) {
          let msg = 'The desired capabilities must include avd';

          _logger.default.errorAndThrow(msg);
        }
      }
    }

    return true;
  }

  setupNewIOSDriver(caps) {
    return (0, _asyncToGenerator2.default)(function* () {
      let iosArgs = {
        javascriptEnabled: true
      };
      let iosdriver = new _appiumIosDriver.default(iosArgs);

      if (caps.platformVersion) {
        let majorVer = caps.platformVersion.toString().split(".")[0];

        if (parseInt(majorVer, 10) >= 10) {
          iosdriver = new _appiumXcuitestDriver.default(iosArgs);
        }
      }

      let capsCopy = _lodash.default.cloneDeep(caps);

      capsCopy.newCommandTimeout = 0;
      yield iosdriver.createSession(capsCopy);
      return iosdriver;
    })();
  }

  startIOSSession(caps) {
    var _this6 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _logger.default.info("Starting an IOS proxy session");

      _this6.proxyAllowList = TO_PROXY_IOS;
      _this6.proxydriver = yield _this6.setupNewIOSDriver(caps);
    })();
  }

  setupNewAndroidDriver(caps) {
    return (0, _asyncToGenerator2.default)(function* () {
      let androidArgs = {
        javascriptEnabled: true
      };
      let androiddriver = new _appiumAndroidDriver.default(androidArgs);

      let capsCopy = _lodash.default.cloneDeep(caps);

      capsCopy.newCommandTimeout = 0;
      yield androiddriver.createSession(capsCopy);
      return androiddriver;
    })();
  }

  startAndroidSession(caps) {
    var _this7 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _logger.default.info("Starting an Android proxy session");

      _this7.proxyAllowList = TO_PROXY_ANDROID;
      _this7.proxydriver = yield _this7.setupNewAndroidDriver(caps);
    })();
  }

  setupNewMacDriver(caps) {
    return (0, _asyncToGenerator2.default)(function* () {
      let macArgs = {
        javascriptEnabled: true
      };
      let macdriver = new _appiumMacDriver.default(macArgs);

      let capsCopy = _lodash.default.cloneDeep(caps);

      capsCopy.newCommandTimeout = 0;
      yield macdriver.createSession(capsCopy);
      return macdriver;
    })();
  }

  startMacSession(caps) {
    var _this8 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _logger.default.info("Starting a Mac proxy session");

      _this8.proxyAllowList = TO_PROXY_MAC;
      _this8.proxydriver = yield _this8.setupNewMacDriver(caps);
    })();
  }

  startYIMacSession(caps) {
    _logger.default.info("Killing app if run");

    let shell = require('shelljs');

    let process_name = caps.app.substring(caps.app.lastIndexOf("/") + 1);
    shell.exec("killall " + process_name);

    _logger.default.info("Launching macOS app");

    this.validateAppLocation(caps.app);

    let spawn = require('child_process').spawn,
        ls = spawn(caps.app);

    ls.stdout.on('data', function (data) {
      if (data != null && caps.showXcodeLog === true) {
        _logger.default.debug('Xcode Log Output: ' + data.toString());
      }
    });
    ls.stderr.on('data', function (data) {
      if (data != null && caps.showXcodeLog === true) {
        _logger.default.debug('Xcode Log Error: ' + data.toString());
      }
    });
    ls.on('exit', function (code) {
      if (code != null && caps.showXcodeLog === true) {
        _logger.default.debug('Application exited with code ' + code.toString());
      }
    });
  }

  endYIMacSession(caps) {
    _logger.default.info("Deleting app");

    let shell = require('shelljs');

    let process_name = caps.app.substring(caps.app.lastIndexOf("/") + 1);
    shell.exec("killall " + process_name);
  }

  startYIRokuSession(caps) {
    _logger.default.info("Launching Roku app");

    this.validateAppLocation(caps.app);

    let shell = require('shelljs');

    let roku_install_script = "curl -v -# -f -i --user '" + caps.username + ":" + caps.password + "' --digest --progress-bar -F 'mysubmit=Install' -F 'archive=@" + caps.app + "' -F 'passwd=' http://" + caps.youiEngineAppAddress + "/plugin_install | grep '<font color' | sed 's/<font color=\'red\'>//' ";
    shell.exec(roku_install_script);
  }

  startYITVOSSession(caps) {
    _logger.default.info("Launching tvOS app");

    this.validateAppLocation(caps.app);

    let shell = require('shelljs');

    if (caps.udid) {
      shell.exec("ios-deploy --id " + caps.udid + " --uninstall --justlaunch --bundle " + caps.app);
    } else {
      shell.exec("ios-deploy --uninstall --justlaunch --bundle " + caps.app);
    }
  }

  endYITVOSSession(caps) {
    _logger.default.info("Deleting app");

    let shell = require('shelljs');

    let bundleid = shell.exec(`osascript -e 'id of app "${caps.app}"'`);

    if (caps.udid) {
      shell.exec("ios-deploy --id " + caps.udid + " --uninstall_only --bundle_id " + bundleid);
    } else {
      shell.exec("ios-deploy --uninstall_only --bundle_id " + bundleid);
    }
  }

  validateAppLocation(app) {
    const fs = require('fs');

    const path = require('path');

    if (!fs.existsSync(app)) {
      let absolutepath = path.resolve(app);
      let msg = 'The app could not be found in following location: ' + absolutepath;

      _logger.default.errorAndThrow(msg);
    }
  }

  connectSocket() {
    var _this9 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let retryCount = 0;
      let connected = false;

      while (retryCount < MAX_RETRY_COUNT && !connected) {
        if (retryCount > 0) {
          _logger.default.info("Waiting " + RETRY_BACKOFF / 1000 + " seconds before trying...");

          yield (0, _asyncbox.sleep)(RETRY_BACKOFF);
        }

        _logger.default.info("Attempt #" + (retryCount + 1));

        let connectedPromise = new _bluebird.default(resolve => {
          let net = require('net');

          let HOST = _this9.opts.youiEngineAppAddress;
          let PORT = 12345;

          _logger.default.info('Connecting to WebDriver: ' + HOST + ':' + PORT);

          _this9.socket = new net.Socket();

          _this9.socket.on('error', function (ex) {
            _logger.default.error(ex);

            _logger.default.error('Check that WebDriver is enabled in application, if a device ensure the proper IP address is used.');

            resolve(false);
          });

          _this9.socket.on('close', function () {
            _logger.default.info('Connection closed');
          });

          _this9.socket.on('timeout', function () {
            _logger.default.error('Connection timed out');

            resolve(false);
          });

          _this9.socket.connect(PORT, HOST, function () {
            _logger.default.info('Connected');

            resolve(true);
          });
        });
        retryCount++;
        connected = yield connectedPromise;

        if (!connected && retryCount === MAX_RETRY_COUNT - 1) {
          _logger.default.errorAndThrow("Failed to connect " + MAX_RETRY_COUNT + " times. Aborting.");
        }
      }

      retryCount = 0;
      _this9.ready = connected;
    })();
  }

  executeSocketCommand(cmd) {
    var _this10 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (!_this10.socket.writable) {
        _logger.default.info("Socket is not writable. Trying to reconnect.");

        yield _this10.connectSocket();
      }

      let cmdPromise = new _bluebird.default(resolve => {
        _logger.default.debug('COMMAND: ' + cmd);

        let totaldata = [];
        let endMarker = new Buffer("youiend");
        let socketClient = _this10.socket;

        let dataHandler = function dataHandler(data) {
          if (data.length >= endMarker.length) {
            let dataend = new Buffer(endMarker.length);
            let startIndex = data.length - endMarker.length;
            data.copy(dataend, 0, startIndex, startIndex + endMarker.length);

            if (dataend.equals(endMarker)) {
              let lastData = data.slice(0, startIndex);
              totaldata.push(lastData);
              socketClient.removeListener('data', dataHandler);
              resolve(Buffer.concat(totaldata));
            } else {
              totaldata.push(data);
            }
          }
        };

        socketClient.write(cmd + "\n", "UTF8", () => {
          socketClient.on('data', dataHandler);
        });
      });
      return yield cmdPromise;
    })();
  }

}

exports.YouiEngineDriver = YouiEngineDriver;
var _iteratorNormalCompletion2 = true;
var _didIteratorError2 = false;
var _iteratorError2 = undefined;

try {
  for (var _iterator2 = _lodash.default.toPairs(_index.default)[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
    let _step2$value = (0, _slicedToArray2.default)(_step2.value, 2),
        cmd = _step2$value[0],
        fn = _step2$value[1];

    YouiEngineDriver.prototype[cmd] = fn;
  }
} catch (err) {
  _didIteratorError2 = true;
  _iteratorError2 = err;
} finally {
  try {
    if (!_iteratorNormalCompletion2 && _iterator2.return != null) {
      _iterator2.return();
    }
  } finally {
    if (_didIteratorError2) {
      throw _iteratorError2;
    }
  }
}require('source-map-support').install();
